core/src/main/java/org/springframework/security/oauth2/server/authorization/client/RegisteredClient.java

```diff
@@ -36,4 +50,369 @@
+				validateClientCredentialsGrantType();
+			}
+			this.validateScopes();
+			this.validateRedirectUris();
+			return this.create();
+		}
+
+		private RegisteredClient create() {
+			RegisteredClient registeredClient = new RegisteredClient();
+
+			registeredClient.id = this.id;
+			registeredClient.clientId = this.clientId;
+			registeredClient.clientSecret = StringUtils.hasText(this.clientSecret) ? this.clientSecret : "";
+			registeredClient.clientAuthenticationMethods = this.clientAuthenticationMethods;
+			if (CollectionUtils.isEmpty(this.clientAuthenticationMethods) && StringUtils.hasText(this.clientSecret)) {
+				registeredClient.clientAuthenticationMethods = Collections.singleton(ClientAuthenticationMethod.BASIC);
+			}
+			registeredClient.authorizationGrantTypes = this.authorizationGrantTypes;
+			registeredClient.redirectUris = this.redirectUris;
+			registeredClient.scopes = this.scopes;
+
+			return registeredClient;
+		}
+
+		private void validateClientCredentialsGrantType() {
+			Assert.isTrue(this.authorizationGrantTypes.contains(AuthorizationGrantType.CLIENT_CREDENTIALS),
+					() -> "authorizationGrantTypes must contain " +
+							AuthorizationGrantType.CLIENT_CREDENTIALS.getValue());
+			Assert.hasText(this.id, "id cannot be empty");
+			Assert.hasText(this.clientId, "clientId cannot be empty");
+		}
+
+		private void validateAuthorizationCodeGrantType() {
+			Assert.isTrue(this.authorizationGrantTypes.contains(AuthorizationGrantType.AUTHORIZATION_CODE),
+					() -> "authorizationGrantTypes must contain " +
+							AuthorizationGrantType.AUTHORIZATION_CODE.getValue());
+			Assert.hasText(this.id, "id cannot be empty");
+			Assert.hasText(this.clientId, "clientId cannot be empty");
+			Assert.notEmpty(this.redirectUris, "redirectUris cannot be empty");
+		}
+
+		private void validateScopes() {
+			if (CollectionUtils.isEmpty(this.scopes)) {
+				return;
+			}
+
+			for (String scope : this.scopes) {
+				Assert.isTrue(validateScope(scope), "scope \"" + scope + "\" contains invalid characters");
+			}
+		}
+
+		private static boolean validateScope(String scope) {
+			return scope == null ||
+					scope.chars().allMatch(c -> withinTheRangeOf(c, 0x21, 0x21) ||
+							withinTheRangeOf(c, 0x23, 0x5B) ||
+							withinTheRangeOf(c, 0x5D, 0x7E));
+		}
+
+		private static boolean withinTheRangeOf(int c, int min, int max) {
+			return c >= min && c <= max;
+		}
+
+		private void validateRedirectUris() {
+			if (CollectionUtils.isEmpty(this.redirectUris)) {
+				return;
+			}
+
+			for (String redirectUri : redirectUris) {
+				Assert.isTrue(validateRedirectUri(redirectUri),
+						"redirect_uri \"" + redirectUri + "\" is not a valid redirect URI or contains fragment");
+			}
+		}
+
+		private static boolean validateRedirectUri(String redirectUri) {
+			try {
+				URI validRedirectUri = new URI(redirectUri);
+				return validRedirectUri.getFragment() == null;
+			} catch (URISyntaxException ex) {
+				return false;
```

Comment:
Could you maybe log the exception here on trace or debug level? This might help in some cases

Reply:
We'll add logging later on